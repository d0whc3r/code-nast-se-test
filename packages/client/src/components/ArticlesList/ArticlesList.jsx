import ArticleCard from '../ArticleCard/ArticleCard'
import Pagination from '../Pagination/Pagination'
import './ArticlesList.styles.css'

function ArticlesList({ articles }) {

  const articleId = (article) => `${article.source.name}-${article.author}-${article.title}`

  return (
    <>
      <section className="list">
        {articles?.map((article) => <ArticleCard key={articleId(article)} article={article} />)}
      </section>
      <Pagination />
    </>
  )
}

export default ArticlesList
