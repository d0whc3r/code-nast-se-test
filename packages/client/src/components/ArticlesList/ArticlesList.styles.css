.list {
    @apply mt-6 grid gap-x-6 gap-y-10 justify-items-center z-10;
    grid-template-columns: repeat(auto-fit, minmax(350px, 1fr));
}
