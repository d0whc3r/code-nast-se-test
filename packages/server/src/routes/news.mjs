import fetch from 'node-fetch'
import { generateUrl } from '../common.mjs'

const PAGE_SIZE = 10

/**
 * Wrapper function to get news and return in http response
 * @param params {Record<string, string>}
 * @param res {import('http').ServerResponse}
 */
async function getNews(params, res) {
  const page = +params.page || 1
  const pageSize = +params.pageSize || PAGE_SIZE
  const url = generateUrl({ ...params, pageSize: PAGE_SIZE })
  return fetch(url)
    .then((response) => response.json())
    .then((data) => {
      if (data.status === 'ok') {
        res.status(200).json({ ...data, pageSize, page, totalPages: Math.ceil(data.totalResults / PAGE_SIZE) })
      } else {
        res.status(500).json({ error: data.message })
      }
    })
    .catch((error) => {
      res.status(500).json({ error })
    })
}

/**
 * Get a list of news by topic
 * @param req {import('http').IncomingMessage}
 * @param res {import('http').ServerResponse}
 * @param next {import('express').NextFunction}
 */
export async function newsByTopic(req, res, next) {
  const params = { ...req.params, ...req.query }
  return getNews(params, res)
}

/**
 * Get a list of news by topic
 * @param req {import('http').IncomingMessage}
 * @param res {import('http').ServerResponse}
 * @param next {import('express').NextFunction}
 */
export async function latestNews(req, res, next) {
  const today = new Date().toISOString().split('T')[0]
  const yesterday = new Date(new Date().setDate(new Date().getDate() - 1)).toISOString().split('T')[0]
  const params = { sources: 'engadget', from: yesterday, to: today, ...req.query }
  return getNews(params, res)
}
