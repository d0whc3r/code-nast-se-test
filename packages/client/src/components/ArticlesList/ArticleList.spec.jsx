import { describe, expect, it } from 'vitest'
import { renderWithPagination } from '../../../test/utils'
import { mockedNews } from '../../api/news.mock'
import ArticlesList from './ArticlesList'

const articles = mockedNews.articles

describe('ArticlesList', () => {
  it('should render', () => {
    const { baseElement } = renderWithPagination(<ArticlesList articles={articles} />)
    expect(baseElement).toMatchSnapshot()
  })
  it('should render without params', () => {
    const { baseElement } = renderWithPagination(<ArticlesList />)
    expect(baseElement).toMatchSnapshot()
  })
})
