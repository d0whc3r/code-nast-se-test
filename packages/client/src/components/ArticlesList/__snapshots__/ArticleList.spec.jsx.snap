// Vitest Snapshot v1

exports[`ArticlesList > should render 1`] = `
<body>
  <div>
    <section
      class="list"
    >
      <a
        class="card"
        href="https://www.enriquedans.com/2022/05/rusia-y-las-vpn-como-sintoma.html"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Rusia y las VPN como síntoma
          </h2>
        </header>
        <img
          alt="Rusia y las VPN como síntoma"
          class="image"
          src="https://www.enriquedans.com/wp-content/uploads/2022/05/VPN.jpeg"
        />
        <p>
          Las descargas y el uso de redes privadas virtuales (VPNs) en Rusia se han disparado desde el inicio de la brutal invasión de Ucrania, con niveles que se aproximan al medio millón de descargas diario … [+3209 chars]
        </p>
        <p
          class="published"
        >
          07/05/2022, 16:46:50
        </p>
      </a>
      <a
        class="card"
        href="https://www.microsiervos.com/archivo/espacio/primer-lanzamiento-orbital-cohete-angara-1-2.html"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Rusia hace el primer lanzamiento orbital de un cohete Angará 1.2 tras 25 años de desarrollo
          </h2>
        </header>
        <img
          alt="Rusia hace el primer lanzamiento orbital de un cohete Angará 1.2 tras 25 años de desarrollo"
          class="image"
          src="https://img.youtube.com/vi/TBEV3foZMKE/maxresdefault.jpg"
        />
        <p>
          Por @Wicho 30 de Abril de 2022
En la noche del 29 de abril de 2022 Rusia hizo por fin el primer lanzamiento orbital de un cohete Angará 1.2 después de unos 25 años de desarrollo en los que hubo vari… [+2292 chars]
        </p>
        <p
          class="published"
        >
          30/04/2022, 13:24:34
        </p>
      </a>
      <a
        class="card"
        href="https://www.genbeta.com/actualidad/este-video-presidente-ucrania-diciendo-que-esnifa-cocaina-bulo-que-ha-usado-tecnologia-para-cambiar-original"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Este vídeo del presidente de Ucrania diciendo que esnifa cocaína es un bulo que ha usado tecnología para cambiar el original
          </h2>
        </header>
        <img
          alt="Este vídeo del presidente de Ucrania diciendo que esnifa cocaína es un bulo que ha usado tecnología para cambiar el original"
          class="image"
          src="https://i.blogs.es/f8fe53/1366_2000/840_560.jpeg"
        />
        <p>
          Ha comenzado a circular un vídeo por redes sociales del presidente de Ucrania, Volodimir Zelenski, afirmando que consume cocaína, que esta sustancia es el mejor energizante y que recomienda su uso a … [+2261 chars]
        </p>
        <p
          class="published"
        >
          27/04/2022, 14:40:57
        </p>
      </a>
      <a
        class="card"
        href="https://hipertextual.com/2022/04/rebrote-covid-19-china-chips"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            El rebrote de la COVID-19 en China golpea con fuerza a la industria de los chips
          </h2>
        </header>
        <img
          alt="El rebrote de la COVID-19 en China golpea con fuerza a la industria de los chips"
          class="image"
          src="https://hipertextual.com/wp-content/uploads/2021/09/cpu-chip-processor-computer.jpeg"
        />
        <p>
          El rebrote de la COVID-19 en China ha provocado la implementación de duras medidas restrictivas en algunas de sus ciudades más importantes. Shanghái es una de las urbes que más las padece, generando … [+3145 chars]
        </p>
        <p
          class="published"
        >
          18/04/2022, 21:49:06
        </p>
      </a>
      <a
        class="card"
        href="https://hipertextual.com/2022/04/the-witcher-3-resta-importancia-nueva-demora-next-gen"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            CD Projekt RED calma las aguas por el nuevo retraso de ‘The Witcher 3’ para PS5 y Xbox Series X|S
          </h2>
        </header>
        <img
          alt="CD Projekt RED calma las aguas por el nuevo retraso de ‘The Witcher 3’ para PS5 y Xbox Series X|S"
          class="image"
          src="https://hipertextual.com/wp-content/uploads/2019/12/hipertextual-nuevos-juegos-the-witcher-horizonte-cd-projekt-red-sapkowski-hacen-paces-2019203367.jpg"
        />
        <p>
          El más reciente retraso que ha sufrido el lanzamiento de The Witcher 3 para PS5 y Xbox Series X|S ha obligado a CD Projekt RED a salir a bajarle el tono a la noticia. En una llamada con inversores (v… [+2933 chars]
        </p>
        <p
          class="published"
        >
          18/04/2022, 23:50:00
        </p>
      </a>
      <a
        class="card"
        href="https://www.genbeta.com/actualidad/vetado-uso-rusia-times-new-roman-arial-verdana-otras-fuentes-culpa-sanciones-no-no-software-libre"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Vetado el uso en Rusia de Times New Roman, Arial, Verdana y otras fuentes, por culpa de las sanciones (no, no son software libre)
          </h2>
        </header>
        <img
          alt="Vetado el uso en Rusia de Times New Roman, Arial, Verdana y otras fuentes, por culpa de las sanciones (no, no son software libre)"
          class="image"
          src="https://i.blogs.es/27cfb9/fuentes/840_560.jpg"
        />
        <p>
          La mayoría de equipos informáticos del mundo, sean Windows, Mac o Linux, incluyen un 'paquete básico' de tipografías preinstaladas, que a todos nos sonarán: Times New Roman, Arial, Helvetica, Verdana… [+3079 chars]
        </p>
        <p
          class="published"
        >
          15/04/2022, 12:29:30
        </p>
      </a>
      <a
        class="card"
        href="https://hipertextual.com/2022/05/canada-condena-crimenes-luna"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Canadá aprueba una ley para condenar los crímenes que ocurran en la Luna
          </h2>
        </header>
        <img
          alt="Canadá aprueba una ley para condenar los crímenes que ocurran en la Luna"
          class="image"
          src="https://i0.wp.com/hipertextual.com/wp-content/uploads/2022/03/astronauta-gravity-espacio.jpg?fit=2200%2C1237&quality=50&strip=all&ssl=1"
        />
        <p>
          Canadá suele ser noticia por muy pocas cosas. Pero en este caso, el tranquilo y apacible país norteamericano salta a la palestra por la aprobación de una ley que permitirá enjuiciar los crímenes come… [+3194 chars]
        </p>
        <p
          class="published"
        >
          06/05/2022, 14:30:00
        </p>
      </a>
      <a
        class="card"
        href="https://www.xataka.com/otros/taiwan-recibe-95-su-internet-cables-submarinos-vulnerables-china-ahora-teme-quedarse-aislada"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Taiwán recibe el 95% de su Internet por cables submarinos vulnerables ante China. Ahora teme quedarse aislada
          </h2>
        </header>
        <img
          alt="Taiwán recibe el 95% de su Internet por cables submarinos vulnerables ante China. Ahora teme quedarse aislada"
          class="image"
          src="https://i.blogs.es/b287cd/1366_2000-1-/840_560.jpg"
        />
        <p>
          La guerra de Ucrania está teniendo un efecto peculiar en el Pacífico, a miles de kilómetros de Kiev: ha avivado el recelo con el que Taiwán ve su frágil conexión a Internet. Más que con la infraestru… [+4821 chars]
        </p>
        <p
          class="published"
        >
          23/04/2022, 20:50:57
        </p>
      </a>
      <a
        class="card"
        href="https://www.xataka.com/empresas-y-economia/netflix-pierde-suscriptores-primera-vez-decada-se-desploma-bolsa"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Netflix pierde suscriptores por primera vez en una década y se desploma en la bolsa
          </h2>
        </header>
        <img
          alt="Netflix pierde suscriptores por primera vez en una década y se desploma en la bolsa"
          class="image"
          src="https://i.blogs.es/7fdcf4/netflix/840_560.jpeg"
        />
        <p>
          Año tras año vimos como Netflix sumaba nuevos usuarios, pese a la creciente competencia, pero esa tendencia se ha detenido. Después del frenazo de 2021, la compañía ha anunciadouna pérdida de 200.000… [+2643 chars]
        </p>
        <p
          class="published"
        >
          20/04/2022, 00:24:45
        </p>
      </a>
      <a
        class="card"
        href="https://www.genbeta.com/actualidad/bulo-reciente-despixelado-bases-rusas-google-maps-permite-detectar-escuadrilla-aviones-gomaespuma"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            El bulo sobre el reciente despixelado de las bases rusas en Google Maps permite detectar una escuadrilla de aviones de gomaespuma
          </h2>
        </header>
        <img
          alt="El bulo sobre el reciente despixelado de las bases rusas en Google Maps permite detectar una escuadrilla de aviones de gomaespuma"
          class="image"
          src="https://i.blogs.es/1e3b4d/rusos/840_560.jpg"
        />
        <p>
          Es de conocimiento público que Google Maps censura habitualmente, desenfocándolas, ciertas instalaciones gubernamentales (sobre todo las de tipo militar) cuando aparecen en sus mapas. El Acuartalamie… [+3198 chars]
        </p>
        <p
          class="published"
        >
          19/04/2022, 21:33:46
        </p>
      </a>
      <a
        class="card"
        href="https://www.genbeta.com/actualidad/criptomonedas-estan-cayendo-enormemente-finales-ano-su-dependencia-a-decisiones-politicas-evidente"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Las criptomonedas están cayendo enormemente desde finales de año. Y su dependencia a las decisiones políticas es evidente
          </h2>
        </header>
        <img
          alt="Las criptomonedas están cayendo enormemente desde finales de año. Y su dependencia a las decisiones políticas es evidente"
          class="image"
          src="https://i.blogs.es/9ee576/1366_2000/840_560.jpeg"
        />
        <p>
          Hay diferentes noticias de los últimos meses que apuntan a que el auge de los criptoactivos podríá haber llegado a su tope. El Banco Central Europeo ha decretado el fin de los tipos de interés a cero… [+3261 chars]
        </p>
        <p
          class="published"
        >
          29/04/2022, 14:23:03
        </p>
      </a>
      <a
        class="card"
        href="https://www.genbeta.com/actualidad/bitcoin-bancarrota-salvador-fue-primer-pais-adoptarlo-ahora-vive-consecuencias-su-gran-caida"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Tras el bitcoin, la bancarrota: El Salvador fue el primer país en adoptarlo y ahora vive las consecuencias de su gran caída
          </h2>
        </header>
        <img
          alt="Tras el bitcoin, la bancarrota: El Salvador fue el primer país en adoptarlo y ahora vive las consecuencias de su gran caída"
          class="image"
          src="https://i.blogs.es/6c59d6/1366_2000/840_560.jpg"
        />
        <p>
          El Salvador fue el primer país del mundo en adoptar el bitcoin como una moneda de curso legal. Hace unos días la República Centroafricana hizo lo mismo, aunque la decisión apunta más a una decisión d… [+2800 chars]
        </p>
        <p
          class="published"
        >
          13/05/2022, 11:46:22
        </p>
      </a>
      <a
        class="card"
        href="https://www.muyinteresante.es/ciencia/articulo/rusia-lanza-una-nave-espacial-militar-secreta-en-orbita-alrededor-de-la-tierra-601651560217"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Rusia lanza una nave espacial militar secreta en órbita alrededor de la Tierra
          </h2>
        </header>
        <img
          alt="Rusia lanza una nave espacial militar secreta en órbita alrededor de la Tierra"
          class="image"
          src="https://estaticos.muyinteresante.es/uploads/images/article/6270cf7a5cafe83ce49b92a1/dealwithit_0.jpg"
        />
        <p>
          Rusia continúa con sus lanzamientos espaciales a pesar de las continuas cancelaciones espaciales (y no espaciales) a nivel internacional por el inicio de la invasión de Ucrania por la Rusia de Vladim… [+1629 chars]
        </p>
        <p
          class="published"
        >
          03/05/2022, 08:51:00
        </p>
      </a>
      <a
        class="card"
        href="https://www.muyinteresante.es/ciencia/articulo/la-esa-no-volvera-a-colaborar-con-rusia-en-futuras-misiones-a-la-luna-511649948346"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            La ESA no volverá a colaborar con Rusia en futuras misiones a la Luna
          </h2>
        </header>
        <img
          alt="La ESA no volverá a colaborar con Rusia en futuras misiones a la Luna"
          class="image"
          src="https://estaticos.muyinteresante.es/uploads/images/article/625836795cafe8f75b22723d/luna-desaparecio.jpg"
        />
        <p>
          La Agencia Espacial Europea ha puesto fin, definitivamente, a su colaboración con Rusia en futuras misiones lunares. No volverán a cooperar con la agencia espacial Roscosmos de Moscú en las misiones … [+1641 chars]
        </p>
        <p
          class="published"
        >
          14/04/2022, 17:08:49
        </p>
      </a>
      <a
        class="card"
        href="https://hipertextual.com/2022/04/se-acabo-el-baile-para-netflix"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            ¿Se acabó el baile para Netflix?
          </h2>
        </header>
        <img
          alt="¿Se acabó el baile para Netflix?"
          class="image"
          src="https://hipertextual.com/wp-content/uploads/2022/02/thibault-penin-awol7qqsffm-unsplash-scaled.jpeg"
        />
        <p>
          Por primera vez en una década, Netflix ha perdido suscriptores durante el primer trimestre de 2022. Concretamente, han sido 200.000. Y, lo más preocupante de todo, es que la compañía proyecta perder … [+9027 chars]
        </p>
        <p
          class="published"
        >
          20/04/2022, 17:43:13
        </p>
      </a>
      <a
        class="card"
        href="https://www.elmundo.es/internacional/2022/05/12/627cfad6fc6c8369208b4584.html"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Rusia avisa de "medidas militares" ante la entrada de Finlandia en la OTAN
          </h2>
        </header>
        <img
          alt="Rusia avisa de \\"medidas militares\\" ante la entrada de Finlandia en la OTAN"
          class="image"
          src="https://phantom-elmundo.unidadeditorial.es/8e76ff989b2fada7368f5b9f244eb5ae/resize/1200/f/jpg/assets/multimedia/imagenes/2022/05/12/16523576029531.jpg"
        />
        <p>
          Rusia se verá obligada a tomar medidas militares y de otra índole ante el ingreso de Suecia y Finlandia en la OTAN, advirtió el Ministerio de Exteriores de Rusia. El Kremlin indicó que la reacción de… [+4115 chars]
        </p>
        <p
          class="published"
        >
          12/05/2022, 15:36:51
        </p>
      </a>
      <a
        class="card"
        href="https://www.xataka.com/energia/reino-unido-quiere-construir-ahora-reactor-nuclear-al-ano-mucho-dificil-que-parece"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Reino Unido quiere construir ahora un reactor nuclear al año. Es mucho más difícil de lo que parece
          </h2>
        </header>
        <img
          alt="Reino Unido quiere construir ahora un reactor nuclear al año. Es mucho más difícil de lo que parece"
          class="image"
          src="https://i.blogs.es/7ff8b7/reino-unido-centrales-nucleares/840_560.jpeg"
        />
        <p>
          La coyuntura actual está cambiando las reglas del juego. Durante los últimos años hemos presenciado cómo algunos países europeos, entre los que se encuentran España y Alemania, apuestan por un modelo… [+5965 chars]
        </p>
        <p
          class="published"
        >
          03/05/2022, 17:01:15
        </p>
      </a>
      <a
        class="card"
        href="https://www.marketingdirecto.com/anunciantes-general/anunciantes/henkel-enesima-multinacional-suspende-operaciones-rusia"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Henkel, la enésima multinacional que suspende sus operaciones en Rusia
          </h2>
        </header>
        <img
          alt="Henkel, la enésima multinacional que suspende sus operaciones en Rusia"
          class="image"
          src="https://www.marketingdirecto.com/wp-content/uploads/2022/04/henkel.jpg"
        />
        <p>
          Henkel ha anunciado este martes su decisión de abandonar Rusia a la luz del actual desarrollo de los acontecimientos en Ucrania.Desde que la guerra de Ucrania estallara hace ya casi dos meses toda un… [+2230 chars]
        </p>
        <p
          class="published"
        >
          19/04/2022, 17:49:17
        </p>
      </a>
      <a
        class="card"
        href="https://www.xataka.com/movilidad/pensabamos-que-gasolina-estaba-cara-europa-quiere-dejar-importar-petroleo-ruso-a-finales-ano"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            Y pensábamos que la gasolina estaba cara: Europa quiere dejar de importar petróleo ruso a finales de año
          </h2>
        </header>
        <img
          alt="Y pensábamos que la gasolina estaba cara: Europa quiere dejar de importar petróleo ruso a finales de año"
          class="image"
          src="https://i.blogs.es/35b498/refineria-buena/840_560.jpg"
        />
        <p>
          La invasión rusa de Ucrania sigue adelante. Después de muchas otras sanciones y de eternas discusiones entre los países que conforman la Unión Europea, la presidenta de la Comisión Europea, Ursula vo… [+3387 chars]
        </p>
        <p
          class="published"
        >
          04/05/2022, 14:03:23
        </p>
      </a>
      <a
        class="card"
        href="https://www.muyinteresante.es/tecnologia/articulo/quien-esta-detras-de-los-ciberataques-a-ucrania-521650353549"
        target="_blank"
      >
        <header>
          <h2
            class="title"
          >
            ¿Quién está detrás de los ciberataques a Ucrania?
          </h2>
        </header>
        <img
          alt="¿Quién está detrás de los ciberataques a Ucrania?"
          class="image"
          src="https://estaticos.muyinteresante.es/uploads/images/article/625e65b45cafe89c7a65a1e4/19.jpg"
        />
        <p>
          Ucrania se ha visto afectada por un aluvión de ataques cibernéticos, que según su gobierno están en un nivel completamente diferente a los que se habían detectado hasta el momento. Hace unos días, an… [+2630 chars]
        </p>
        <p
          class="published"
        >
          19/04/2022, 09:41:10
        </p>
      </a>
    </section>
    <div
      class="container"
    >
      <button
        class="button-text"
        disabled=""
      >
        First
      </button>
      <button
        class="button-text"
        disabled=""
      >
        Previous
      </button>
      <button
        class="button-text"
        disabled=""
      >
        Next
      </button>
      <button
        class="button-text"
        disabled=""
      >
        Last
      </button>
    </div>
  </div>
</body>
`;

exports[`ArticlesList > should render without params 1`] = `
<body>
  <div>
    <section
      class="list"
    />
    <div
      class="container"
    >
      <button
        class="button-text"
        disabled=""
      >
        First
      </button>
      <button
        class="button-text"
        disabled=""
      >
        Previous
      </button>
      <button
        class="button-text"
        disabled=""
      >
        Next
      </button>
      <button
        class="button-text"
        disabled=""
      >
        Last
      </button>
    </div>
  </div>
</body>
`;
